### PR TITLE
Update VH Leptonic Tags under STXS workflow 

### PR DESCRIPTION
**Pull request by youyingli**
- Update WH Leptonic Tag with full RunII MVA training
- VHLeptonicLooseTag is absorbed into WH Leptonic Tag
- Split WH Leptonic Tag into several reco categories according to MVA and diphoton pt
- Update ZH Leptonic Tag with cut-based selection

### CODE REVIEW

File: Taggers/plugins/WHLeptonicTagProducer.cc [hunk 1]
```diff
@@ -45,366 +46,463 @@ namespace flashgg {
...
-            std::vector<edm::Ptr<flashgg::Muon> > goodMuons = selectMuons( theMuons->ptrs(), dipho, vertices->ptrs(), muonEtaThreshold_, leptonPtThreshold_,
-                    muPFIsoSumRelThreshold_, deltaRMuonPhoThreshold_, deltaRMuonPhoThreshold_ );
-            
-            std::vector<edm::Ptr<Electron> >goodElectrons = selectStdElectrons( theElectrons->ptrs(), dipho,vertices->ptrs(), leptonPtThreshold_, electronEtaThresholds_,
-                                                                                useElectronMVARecipe_,useElectronLooseID_,
-                                                                                deltaRPhoElectronThreshold_,DeltaRTrkElec_,deltaMassElectronZThreshold_,
-                                                                                rho_, evt.isRealData() );
-            
-            hasGoodElec = ( goodElectrons.size() > 0 );
+            std::vector<edm::Ptr<Electron> >goodElectrons = selectStdElectrons( theElectrons->ptrs(),
```
Comment (simonepigazzini) [inline Taggers/plugins/WHLeptonicTagProducer.cc:398]: From which file are you loading this? please check the issue here: 
https://github.com/cms-analysis/flashgg/issues/1195

File: Taggers/python/flashggTagSorter_cfi.py [hunk 2]
```diff
@@ -20,7 +20,7 @@
 #        cms.PSet(TagName = cms.InputTag('flashggVHLooseTag')),
         cms.PSet(TagName = cms.InputTag('flashggZHLeptonicTag')),
         cms.PSet(TagName = cms.InputTag('flashggWHLeptonicTag')),
-        cms.PSet(TagName = cms.InputTag('flashggVHLeptonicLooseTag')),
+#        cms.PSet(TagName = cms.InputTag('flashggVHLeptonicLooseTag')),
```
Comment (simonepigazzini) [inline Taggers/python/flashggTagSorter_cfi.py]: please cleaup up everything related to VH that is no longer in use (comments will create confusion in the future)

File: Taggers/python/flashggTags_cff.py [hunk 3]
```diff
@@ -435,87 +435,76 @@
...
+                                    SystLabel                   = cms.string(""),
+                                    ElectronTag                 = cms.InputTag('flashggSelectedElectrons'),
+                                    MuonTag                     = cms.InputTag('flashggSelectedMuons'),
+                                    inputTagJets                = UnpackedJetCollectionVInputTag,
+                                    METTag                      = cms.InputTag('flashggMets'),
+                                    VertexTag                   = cms.InputTag('offlineSlimmedPrimaryVertices'),
+                                    rhoTag                      = cms.InputTag('fixedGridRhoFastjetAll'),
+                                    MVAResultTag                = cms.InputTag('flashggDiPhotonMVA'),
+                                    GenParticleTag              = cms.InputTag( "flashggPrunedGenParticles" ),
+                                    WHMVAweightfile             = cms.FileInPath("flashgg/Taggers/data/TMVA_WHLeptonicTag_BDT_16Dec2019.weights.xml"),
```
Comment (simonepigazzini) [inline Taggers/python/flashggTags_cff.py:486]: Is this file common to all years? Is it already on EOS?

**Comment by simonepigazzini** [comment]
Looks good a part for few things I pointed out line by line. Please also check with @edjtscott if the tag producers behavior is the one expected now after the recent changes (I suspect you can remove the TagTruth part).

simone

File: Taggers/plugins/WHLeptonicTagProducer.cc (hunk 1 above)
Comment (youyingli) [reply Taggers/plugins/WHLeptonicTagProducer.cc:398]: Leptonselection.cc is applied in WH Leptonic Tag and ZH Leptonic Tag.

File: Taggers/python/flashggTags_cff.py (hunk 3 above)
Comment (youyingli) [reply Taggers/python/flashggTags_cff.py:486]: Yes, it's for full runII and is put in eos.

File: Taggers/python/flashggTagSorter_cfi.py (hunk 2 above)
Comment (youyingli) [reply Taggers/python/flashggTagSorter_cfi.py]: OK. I have removed VH tags which are not used.

File: Taggers/plugins/WHLeptonicTagProducer.cc (hunk 1 above)
Comment (simonepigazzini) [reply Taggers/plugins/WHLeptonicTagProducer.cc:398]: Are you sure those are the recommended IDs from the respective POGs?

Comment (youyingli) [reply Taggers/plugins/WHLeptonicTagProducer.cc:398]: In LeptonSelection.cc, WH and ZH leptonic tag use selectStdElectrons function which has cut-based loose electron ID called from MIcroAOD (like electron->passLooseId(), I think this is recommended) and photon veto, but LeptonSelection2018.cc uses electron MVA WP so I didn't use it. Maybe these two files should be merged to a single file which can choose cut-based ID or MVA WP.

**Review by edjtscott — COMMENTED** [review]
After a quick look from my side this looks good. Regarding the TagTruth, it's fine to have it in here. The TagSorter now looks for an existing TagTruth, and adds the STXS & truth vertex info if it exists, creating a new TagTruth otherwise.

**Comment by youyingli** [comment]
STXS information now has been added in TagTruth of WH and ZH leptonic Tag

**Comment by edjtscott** [comment]
hi @youyingli and @simonepigazzini - the developments from this PR are now included in [PR 1204](https://github.com/cms-analysis/flashgg/pull/1204) so I will close this one.

**Pull request closed without merging**